[16.0] shopfloor: checkout scan change lot - port of #712

jbaudoux opened the pull request:
Port of:
* https://github.com/OCA/wms/pull/712

cc @sebalix

sebalix committed:
shopfloor, checkout: use lot_not_found_in_picking msg

sebalix committed:
shopfloor, checkout: allow change lot on line select

When it comes to select a line, scanning the wrong lot will invite the
user to scan it again to change it on the relevant line only if there is
one sharing the same product.

simahawk committed:
shopfloor checkout: fix change lot

simahawk committed:
shopfloor checkout: add hook to select lines by lot

sebalix committed:
shopfloor, checkout: check lot on change confirmation

sebalix commented on shopfloor/services/checkout.py:
```diff
@@ -1749,6 +1826,7 @@ def _schema_stock_picking_details(self):
             group_lines_by_location={"type": "boolean"},
             show_oneline_package_content={"type": "boolean"},
             need_confirm_pack_all={"type": "string"},
+            need_confirm_lot={"type": "integer"},
```
The v14 PR was adding `nullable: True` here, removed on purpose?

EDIT: ok you switched from None to False everywhere

sebalix replied:
I do not remember why it was put as `integer+nullable` instead of boolean... like if there was a trap there.

EDIT: was done in this commit: https://github.com/OCA/wms/pull/712/commits/b55b89a648b1958f9d34413edcac1756e46b08c8 which hasn't been ported

jbaudoux replied:
This explains the errors I was getting. I included that missing commit and reduced my mig commit to its minimum

sebalix commented on shopfloor/tests/test_checkout_scan_line.py:
```diff
@@ -248,17 +251,51 @@ def test_scan_line_error_product_not_in_picking(self):
+                "confirm_lot": lot.id,
+            },
+        )
+        message = self.msg_store.lot_replaced_by_lot(previous_lot, lot)
```
Can you explain me why we do not have anymore a draft inventory generated here?

jbaudoux replied:
You don't have inventory in odoo v16 anymore. You can only flag the quant or location to be inventoried without more detail

sebalix reviewed (approved):
LG (compared with v14 PR)

jbaudoux committed:
[MIG] shopfloor checkout: change lot

Adapt to v16
Changing lot requires the lot to exist on the location, you cannot create negative quant anymore

jbaudoux commented:
/ocabot merge minor
